packages/web3-eth-abi/src/types.ts

```diff
@@ -1,57 +1,187 @@
-// TODO: Adding reference of source definition/doc for these types
-export interface JsonAbiParameter {
-	readonly name: string;
-	readonly type: string;
-	readonly baseType?: string;
-	readonly indexed?: boolean;
-	readonly components?: Array<JsonAbiParameter>;
-	readonly arrayLength?: number;
-	readonly arrayChildren?: Array<JsonAbiParameter>;
-}
+import {
+	Address,
+	Bytes,
+	Numbers,
+	ObjectValueToTuple,
+	ArrayToIndexObject,
+	FixedSizeArray,
+} from 'web3-utils';
+import { ConvertToNumber } from './number_map_type';
 
-export interface JsonAbiStruct {
+export interface AbiStruct {
 	[key: string]: unknown;
 	name?: string;
 	type: string;
 }
 
-export interface JsonAbiCoderStruct extends JsonAbiStruct {
+export interface AbiCoderStruct extends AbiStruct {
 	[key: string]: unknown;
-	components?: Array<JsonAbiStruct>;
+	components?: Array<AbiStruct>;
 }
 
+// https://docs.soliditylang.org/en/latest/abi-spec.html#json
+export type AbiParameter = {
+	readonly name: string;
+	readonly type: string;
+	readonly baseType?: string;
+	readonly indexed?: boolean;
+	readonly components?: ReadonlyArray<AbiParameter>;
+	readonly arrayLength?: number;
+	readonly arrayChildren?: ReadonlyArray<AbiParameter>;
+};
+
 type FragmentTypes = 'constructor' | 'event' | 'function';
 
-export interface JsonAbiBaseFragment {
-	name?: string;
-	type: FragmentTypes;
-	inputs?: Array<JsonAbiParameter>;
-}
+export type AbiBaseFragment = {
+	readonly type: FragmentTypes;
+};
 
-export interface JsonAbiConstructorFragment extends JsonAbiBaseFragment {
-	type: 'constructor';
-	stateMutability: 'nonpayable' | 'payable';
-}
+// https://docs.soliditylang.org/en/latest/abi-spec.html#json
+export type AbiConstructorFragment = AbiBaseFragment & {
+	readonly type: 'constructor';
+	readonly stateMutability: 'nonpayable' | 'payable';
+	readonly inputs: ReadonlyArray<AbiParameter>;
+};
 
-export interface JsonAbiFunctionFragment extends JsonAbiBaseFragment {
-	type: 'function';
-	stateMutability: 'nonpayable' | 'payable' | 'pure' | 'view';
-	outputs?: Array<JsonAbiParameter>;
+// https://docs.soliditylang.org/en/latest/abi-spec.html#json
+export type AbiFunctionFragment = AbiBaseFragment & {
+	readonly name: string;
+	readonly type: 'function';
+	readonly stateMutability: 'nonpayable' | 'payable' | 'pure' | 'view';
+	readonly inputs: ReadonlyArray<AbiParameter>;
+	readonly outputs: ReadonlyArray<AbiParameter>;
 
 	// legacy properties
-	constant?: boolean;	// stateMutability == 'pure' or stateMutability == 'view'
-	payable?: boolean; // stateMutability == 'payable'
-}
+	readonly constant?: boolean; // stateMutability == 'pure' or stateMutability == 'view'
```

Comment:
constants and payable are not mentioned in specs, we only have `stateMutability` field for this

Reply:
wts the min version of solidity being type validated with current functions? will be useful to mention min version in code as comments and also in doc.

Reply:
I updated and mentioned in README file. I tested it personally for the ABIs generated by `v0.4.x` solidity. In future if required we can add more tests for different versions of solidity.